File: template/skills/devops/scripts/agentDevopsInventory.ts
// agentDevopsInventory.ts
// Agent DevOps Inventory v1 - TypeScript Parser/Validator
// See: https://github.com/jezweb/claude-skills/skills/server-admin

export type ProviderType =
  | "oci"
  | "hetzner"
  | "gcp"
  | "contabo"
  | "digitalocean"
  | "vultr"
  | "linode"
  | "local_network"
  | "other"
  | string;

export type ConnectVia = "local" | "ssh" | string;

export interface Provider {
  name: string; // e.g. "OCI"
  type?: ProviderType;
  authMethod?: string;
  authFile?: string;
  defaultRegion?: string;
  label?: string;
  notes?: string;
  // Additional arbitrary fields:
  [key: string]: string | undefined;
}

export interface Server {
  id: string; // e.g. "WEB01"
  provider?: string; // e.g. "OCI"
  kind?: string; // vm | physical | local_pc | ...
  name?: string;
  connectVia?: ConnectVia;
  env?: string;
  os?: string;
  role?: string;
  status?: string;
  tags?: string[]; // parsed from comma-separated list
  host?: string;
  port?: number;
  user?: string;
  sshKeyPath?: string;
  notes?: string;
  // Additional arbitrary fields:
  [key: string]: string | string[] | number | undefined;
}

export interface InventoryMetadata {
  version: string; // default "1" if missing
  project?: string;
  owner?: string;
  notes?: string;
  // raw map of other AGENT_DEVOPS_* keys:
  [key: string]: string | undefined;
}

export interface Inventory {
  metadata: InventoryMetadata;
  providers: Record<string, Provider>;
  servers: Record<string, Server>;
}

export interface ValidationError {
  scope: "provider" | "server" | "file";
  id?: string;
  message: string;
}

export interface ParseResult {
  inventory: Inventory;
  errors: ValidationError[];
}

/**
 * Parse a .env-style text into a raw key-value object.
 */
export function parseEnvText(text: string): Record<string, string> {
  const result: Record<string, string> = {};

  const lines = text.split(/\r?\n/);
  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;

    const eqIndex = line.indexOf("=");
    if (eqIndex === -1) {
      // ignore malformed lines for now
      continue;
    }

    const key = line.slice(0, eqIndex).trim();
    const value = line.slice(eqIndex + 1).trim();
    if (!key) continue;

    result[key] = value;
  }

  return result;
}

/**
 * Build Inventory from a key-value map.
 */
export function buildInventory(env: Record<string, string>): ParseResult {
  const providers: Record<string, Provider> = {};
  const servers: Record<string, Server> = {};
  const errors: ValidationError[] = [];

  // Metadata
  const version = env["AGENT_DEVOPS_VERSION"] || "1";

  const metadata: InventoryMetadata = {
    version,
    project: env["AGENT_DEVOPS_PROJECT"],
    owner: env["AGENT_DEVOPS_OWNER"],
    notes: env["AGENT_DEVOPS_NOTES"],
  };

  // include other AGENT_DEVOPS_* keys
  for (const [key, value] of Object.entries(env)) {
    if (key.startsWith("AGENT_DEVOPS_") && !(key in metadata)) {
      (metadata as Record<string, string | undefined>)[key] = value;
    }
  }

  // Providers
  for (const [key, value] of Object.entries(env)) {
    if (!key.startsWith("PROVIDER_")) continue;

    const parts = key.split("_");
    if (parts.length < 3) continue; // malformed: PROVIDER_<NAME>_<FIELD>

    const name = parts[1]; // e.g. OCI
    const field = parts.slice(2).join("_"); // e.g. TYPE, DEFAULT_REGION

    if (!providers[name]) {
      providers[name] = { name };
    }

    const provider = providers[name];

    switch (field) {
      case "TYPE":
        provider.type = value as ProviderType;
        break;
      case "AUTH_METHOD":
        provider.authMethod = value;
        break;
      case "AUTH_FILE":
        provider.authFile = value;
        break;
      case "DEFAULT_REGION":
        provider.defaultRegion = value;
        break;
      case "LABEL":
        provider.label = value;
        break;
      case "NOTES":
        provider.notes = value;
        break;
      default:
        provider[field] = value;
        break;
    }
  }

  // Servers
  for (const [key, value] of Object.entries(env)) {
    if (!key.startsWith("SERVER_")) continue;

    const parts = key.split("_");
    if (parts.length < 3) continue; // malformed: SERVER_<ID>_<FIELD>

    const id = parts[1]; // e.g. WEB01
    const field = parts.slice(2).join("_"); // e.g. HOST, SSH_KEY_PATH

    if (!servers[id]) {
      servers[id] = { id };
    }

    const server = servers[id];

    switch (field) {
      case "PROVIDER":
        server.provider = value;
        break;
      case "KIND":
        server.kind = value;
        break;
      case "NAME":
        server.name = value;
        break;
      case "CONNECT_VIA":
        server.connectVia = value as ConnectVia;
        break;
      case "ENV":
        server.env = value;
        break;
      case "OS":
        server.os = value;
        break;
      case "ROLE":
        server.role = value;
        break;
      case "STATUS":
        server.status = value;
        break;
      case "TAGS":
        server.tags = value
          .split(",")
          .map((t) => t.trim())
          .filter(Boolean);
        break;
      case "HOST":
        server.host = value;
        break;
      case "PORT": {
        const n = Number(value);
        if (!Number.isNaN(n)) {
          server.port = n;
        } else {
          errors.push({
            scope: "server",
            id,
            message: `Invalid PORT for server ${id}: "${value}"`,
          });
        }
        break;
      }
      case "USER":
        server.user = value;
        break;
      case "SSH_KEY_PATH":
        server.sshKeyPath = value;
        break;
      case "NOTES":
        server.notes = value;
        break;
      default:
        // Keep unknown fields as strings
        (server as Record<string, unknown>)[field] = value;
        break;
    }
  }

  // Basic validation
  for (const [name, provider] of Object.entries(providers)) {
    if (!provider.type) {
      errors.push({
        scope: "provider",
        id: name,
        message: `Provider ${name} is missing TYPE`,
      });
    }
  }

  for (const [id, server] of Object.entries(servers)) {
    if (!server.provider) {
      errors.push({
        scope: "server",
        id,
        message: `Server ${id} is missing PROVIDER`,
      });
    }
    if (!server.kind) {
      errors.push({
        scope: "server",
        id,
        message: `Server ${id} is missing KIND`,
      });
    }
    if (!server.name) {
      errors.push({
        scope: "server",
        id,
        message: `Server ${id} is missing NAME`,
      });
    }
    if (!server.connectVia) {
      errors.push({
        scope: "server",
        id,
        message: `Server ${id} is missing CONNECT_VIA`,
      });
    }
  }

  const inventory: Inventory = {
    metadata,
    providers,
    servers,
  };

  return { inventory, errors };
}

/**
 * Convenience function: parse inventory from a .env-style string.
 */
export function parseInventoryFromEnv(text: string): ParseResult {
  const env = parseEnvText(text);
  return buildInventory(env);
}

/**
 * Get servers matching specified criteria.
 */
export function findServers(
  inventory: Inventory,
  criteria: {
    env?: string;
    role?: string;
    provider?: string;
    status?: string;
    tag?: string;
  }
): Server[] {
  return Object.values(inventory.servers).filter((server) => {
    if (criteria.env && server.env !== criteria.env) return false;
    if (criteria.role && server.role !== criteria.role) return false;
    if (criteria.provider && server.provider !== criteria.provider) return false;
    if (criteria.status && server.status !== criteria.status) return false;
    if (criteria.tag && (!server.tags || !server.tags.includes(criteria.tag)))
      return false;
    return true;
  });
}

/**
 * Generate inventory text from Inventory object.
 */
export function serializeInventory(inventory: Inventory): string {
  const lines: string[] = [];

  // Metadata
  lines.push("# =========================");
  lines.push("# METADATA");
  lines.push("# =========================");
  lines.push(`AGENT_DEVOPS_VERSION=${inventory.metadata.version}`);
  if (inventory.metadata.project) {
    lines.push(`AGENT_DEVOPS_PROJECT=${inventory.metadata.project}`);
  }
  if (inventory.metadata.owner) {
    lines.push(`AGENT_DEVOPS_OWNER=${inventory.metadata.owner}`);
  }
  if (inventory.metadata.notes) {
    lines.push(`AGENT_DEVOPS_NOTES=${inventory.metadata.notes}`);
  }
  lines.push("");

  // Providers
  lines.push("# =========================");
  lines.push("# PROVIDERS");
  lines.push("# =========================");
  for (const provider of Object.values(inventory.providers)) {
    lines.push("");
    lines.push(`# ${provider.label || provider.name}`);
    lines.push(`PROVIDER_${provider.name}_TYPE=${provider.type || "other"}`);
    if (provider.authMethod) {
      lines.push(`PROVIDER_${provider.name}_AUTH_METHOD=${provider.authMethod}`);
    }
    if (provider.authFile) {
      lines.push(`PROVIDER_${provider.name}_AUTH_FILE=${provider.authFile}`);
    }
    if (provider.defaultRegion) {
      lines.push(
        `PROVIDER_${provider.name}_DEFAULT_REGION=${provider.defaultRegion}`
      );
    }
    if (provider.label) {
      lines.push(`PROVIDER_${provider.name}_LABEL=${provider.label}`);
    }
    if (provider.notes) {
      lines.push(`PROVIDER_${provider.name}_NOTES=${provider.notes}`);
    }
  }
  lines.push("");

  // Servers
  lines.push("# =========================");
  lines.push("# SERVERS / NODES");
  lines.push("# =========================");
  for (const server of Object.values(inventory.servers)) {
    lines.push("");
    lines.push(`# ${server.name || server.id}`);
    lines.push(`SERVER_${server.id}_PROVIDER=${server.provider || ""}`);
    lines.push(`SERVER_${server.id}_KIND=${server.kind || ""}`);
    lines.push(`SERVER_${server.id}_NAME=${server.name || ""}`);
    lines.push(`SERVER_${server.id}_CONNECT_VIA=${server.connectVia || ""}`);
    if (server.host) lines.push(`SERVER_${server.id}_HOST=${server.host}`);
    if (server.port) lines.push(`SERVER_${server.id}_PORT=${server.port}`);
    if (server.user) lines.push(`SERVER_${server.id}_USER=${server.user}`);
    if (server.sshKeyPath)
      lines.push(`SERVER_${server.id}_SSH_KEY_PATH=${server.sshKeyPath}`);
    if (server.env) lines.push(`SERVER_${server.id}_ENV=${server.env}`);
    if (server.os) lines.push(`SERVER_${server.id}_OS=${server.os}`);
    if (server.role) lines.push(`SERVER_${server.id}_ROLE=${server.role}`);
    if (server.status)
      lines.push(`SERVER_${server.id}_STATUS=${server.status}`);
    if (server.tags && server.tags.length > 0)
      lines.push(`SERVER_${server.id}_TAGS=${server.tags.join(",")}`);
    if (server.notes) lines.push(`SERVER_${server.id}_NOTES=${server.notes}`);
  }

  return lines.join("\n");
}

// Example usage:
// import { parseInventoryFromEnv } from "./agentDevopsInventory";
// import { readFileSync } from "fs";
//
// const text = readFileSync(".agent-devops.env", "utf8");
// const { inventory, errors } = parseInventoryFromEnv(text);
//
// if (errors.length) {
//   console.error("Inventory validation errors:", errors);
// }
//
// console.log("Servers:", Object.keys(inventory.servers));

File: template/skills/devops/scripts/agent_devops_inventory.py
# agent_devops_inventory.py
# Agent DevOps Inventory v1 - Python Parser/Validator
# See: https://github.com/jezweb/claude-skills/skills/server-admin

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple, Any


ProviderType = str  # e.g. "oci", "hetzner", "local_network"
ConnectVia = str    # e.g. "local", "ssh"


@dataclass
class Provider:
    name: str
    type: Optional[ProviderType] = None
    auth_method: Optional[str] = None
    auth_file: Optional[str] = None
    default_region: Optional[str] = None
    label: Optional[str] = None
    notes: Optional[str] = None
    extra: Dict[str, str] = field(default_factory=dict)


@dataclass
class Server:
    id: str
    provider: Optional[str] = None
    kind: Optional[str] = None
    name: Optional[str] = None
    connect_via: Optional[ConnectVia] = None
    env: Optional[str] = None
    os: Optional[str] = None
    role: Optional[str] = None
    status: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    host: Optional[str] = None
    port: Optional[int] = None
    user: Optional[str] = None
    ssh_key_path: Optional[str] = None
    notes: Optional[str] = None
    extra: Dict[str, Any] = field(default_factory=dict)


@dataclass
class InventoryMetadata:
    version: str = "1"
    project: Optional[str] = None
    owner: Optional[str] = None
    notes: Optional[str] = None
    extra: Dict[str, str] = field(default_factory=dict)


@dataclass
class Inventory:
    metadata: InventoryMetadata
    providers: Dict[str, Provider]
    servers: Dict[str, Server]


@dataclass
class ValidationError:
    scope: str  # "provider" | "server" | "file"
    id: Optional[str]
    message: str


def parse_env_text(text: str) -> Dict[str, str]:
    """
    Parse a .env-style text into a dict of key -> value.
    """
    env: Dict[str, str] = {}

    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue

        if "=" not in line:
            # ignore malformed lines for now
            continue

        key, value = line.split("=", 1)
        key = key.strip()
        value = value.strip()

        if not key:
            continue

        env[key] = value

    return env


def build_inventory(env: Dict[str, str]) -> Tuple[Inventory, List[ValidationError]]:
    providers: Dict[str, Provider] = {}
    servers: Dict[str, Server] = {}
    errors: List[ValidationError] = []

    version = env.get("AGENT_DEVOPS_VERSION", "1")
    metadata = InventoryMetadata(
        version=version,
        project=env.get("AGENT_DEVOPS_PROJECT"),
        owner=env.get("AGENT_DEVOPS_OWNER"),
        notes=env.get("AGENT_DEVOPS_NOTES"),
    )

    # Extra metadata keys
    for key, value in env.items():
        if key.startswith("AGENT_DEVOPS_") and key not in {
            "AGENT_DEVOPS_VERSION",
            "AGENT_DEVOPS_PROJECT",
            "AGENT_DEVOPS_OWNER",
            "AGENT_DEVOPS_NOTES",
        }:
            metadata.extra[key] = value

    # Providers
    for key, value in env.items():
        if not key.startswith("PROVIDER_"):
            continue
        parts = key.split("_")
        if len(parts) < 3:
            continue  # malformed

        name = parts[1]
        field_name = "_".join(parts[2:])

        provider = providers.get(name)
        if provider is None:
            provider = Provider(name=name)
            providers[name] = provider

        if field_name == "TYPE":
            provider.type = value
        elif field_name == "AUTH_METHOD":
            provider.auth_method = value
        elif field_name == "AUTH_FILE":
            provider.auth_file = value
        elif field_name == "DEFAULT_REGION":
            provider.default_region = value
        elif field_name == "LABEL":
            provider.label = value
        elif field_name == "NOTES":
            provider.notes = value
        else:
            provider.extra[field_name] = value

    # Servers
    for key, value in env.items():
        if not key.startswith("SERVER_"):
            continue
        parts = key.split("_")
        if len(parts) < 3:
            continue  # malformed

        server_id = parts[1]
        field_name = "_".join(parts[2:])

        server = servers.get(server_id)
        if server is None:
            server = Server(id=server_id)
            servers[server_id] = server

        if field_name == "PROVIDER":
            server.provider = value
        elif field_name == "KIND":
            server.kind = value
        elif field_name == "NAME":
            server.name = value
        elif field_name == "CONNECT_VIA":
            server.connect_via = value
        elif field_name == "ENV":
            server.env = value
        elif field_name == "OS":
            server.os = value
        elif field_name == "ROLE":
            server.role = value
        elif field_name == "STATUS":
            server.status = value
        elif field_name == "TAGS":
            tags = [t.strip() for t in value.split(",") if t.strip()]
            server.tags = tags
        elif field_name == "HOST":
            server.host = value
        elif field_name == "PORT":
            try:
                server.port = int(value)
            except ValueError:
                errors.append(
                    ValidationError(
                        scope="server",
                        id=server_id,
                        message=f"Invalid PORT for server {server_id}: {value!r}",
                    )
                )
        elif field_name == "USER":
            server.user = value
        elif field_name == "SSH_KEY_PATH":
            server.ssh_key_path = value
        elif field_name == "NOTES":
            server.notes = value
        else:
            server.extra[field_name] = value

    # Basic validation
    for name, provider in providers.items():
        if provider.type is None:
            errors.append(
                ValidationError(
                    scope="provider",
                    id=name,
                    message=f"Provider {name} is missing TYPE",
                )
            )

    for sid, server in servers.items():
        if server.provider is None:
            errors.append(
                ValidationError(
                    scope="server",
                    id=sid,
                    message=f"Server {sid} is missing PROVIDER",
                )
            )
        if server.kind is None:
            errors.append(
                ValidationError(
                    scope="server",
                    id=sid,
                    message=f"Server {sid} is missing KIND",
                )
            )
        if server.name is None:
            errors.append(
                ValidationError(
                    scope="server",
                    id=sid,
                    message=f"Server {sid} is missing NAME",
                )
            )
        if server.connect_via is None:
            errors.append(
                ValidationError(
                    scope="server",
                    id=sid,
                    message=f"Server {sid} is missing CONNECT_VIA",
                )
            )

    inventory = Inventory(
        metadata=metadata,
        providers=providers,
        servers=servers,
    )

    return inventory, errors


def parse_inventory_from_env(text: str) -> Tuple[Inventory, List[ValidationError]]:
    """
    Convenience function: parse inventory from a .env-style string.
    """
    env = parse_env_text(text)
    return build_inventory(env)


def find_servers(
    inventory: Inventory,
    env: Optional[str] = None,
    role: Optional[str] = None,
    provider: Optional[str] = None,
    status: Optional[str] = None,
    tag: Optional[str] = None,
) -> List[Server]:
    """
    Get servers matching specified criteria.
    """
    results = []
    for server in inventory.servers.values():
        if env and server.env != env:
            continue
        if role and server.role != role:
            continue
        if provider and server.provider != provider:
            continue
        if status and server.status != status:
            continue
        if tag and (not server.tags or tag not in server.tags):
            continue
        results.append(server)
    return results


def serialize_inventory(inventory: Inventory) -> str:
    """
    Generate inventory text from Inventory object.
    """
    lines: List[str] = []

    # Metadata
    lines.append("# =========================")
    lines.append("# METADATA")
    lines.append("# =========================")
    lines.append(f"AGENT_DEVOPS_VERSION={inventory.metadata.version}")
    if inventory.metadata.project:
        lines.append(f"AGENT_DEVOPS_PROJECT={inventory.metadata.project}")
    if inventory.metadata.owner:
        lines.append(f"AGENT_DEVOPS_OWNER={inventory.metadata.owner}")
    if inventory.metadata.notes:
        lines.append(f"AGENT_DEVOPS_NOTES={inventory.metadata.notes}")
    lines.append("")

    # Providers
    lines.append("# =========================")
    lines.append("# PROVIDERS")
    lines.append("# =========================")
    for provider in inventory.providers.values():
        lines.append("")
        lines.append(f"# {provider.label or provider.name}")
        lines.append(f"PROVIDER_{provider.name}_TYPE={provider.type or 'other'}")
        if provider.auth_method:
            lines.append(f"PROVIDER_{provider.name}_AUTH_METHOD={provider.auth_method}")
        if provider.auth_file:
            lines.append(f"PROVIDER_{provider.name}_AUTH_FILE={provider.auth_file}")
        if provider.default_region:
            lines.append(f"PROVIDER_{provider.name}_DEFAULT_REGION={provider.default_region}")
        if provider.label:
            lines.append(f"PROVIDER_{provider.name}_LABEL={provider.label}")
        if provider.notes:
            lines.append(f"PROVIDER_{provider.name}_NOTES={provider.notes}")
    lines.append("")

    # Servers
    lines.append("# =========================")
    lines.append("# SERVERS / NODES")
    lines.append("# =========================")
    for server in inventory.servers.values():
        lines.append("")
        lines.append(f"# {server.name or server.id}")
        lines.append(f"SERVER_{server.id}_PROVIDER={server.provider or ''}")
        lines.append(f"SERVER_{server.id}_KIND={server.kind or ''}")
        lines.append(f"SERVER_{server.id}_NAME={server.name or ''}")
        lines.append(f"SERVER_{server.id}_CONNECT_VIA={server.connect_via or ''}")
        if server.host:
            lines.append(f"SERVER_{server.id}_HOST={server.host}")
        if server.port:
            lines.append(f"SERVER_{server.id}_PORT={server.port}")
        if server.user:
            lines.append(f"SERVER_{server.id}_USER={server.user}")
        if server.ssh_key_path:
            lines.append(f"SERVER_{server.id}_SSH_KEY_PATH={server.ssh_key_path}")
        if server.env:
            lines.append(f"SERVER_{server.id}_ENV={server.env}")
        if server.os:
            lines.append(f"SERVER_{server.id}_OS={server.os}")
        if server.role:
            lines.append(f"SERVER_{server.id}_ROLE={server.role}")
        if server.status:
            lines.append(f"SERVER_{server.id}_STATUS={server.status}")
        if server.tags:
            lines.append(f"SERVER_{server.id}_TAGS={','.join(server.tags)}")
        if server.notes:
            lines.append(f"SERVER_{server.id}_NOTES={server.notes}")

    return "\n".join(lines)


# Example usage:
# from agent_devops_inventory import parse_inventory_from_env
#
# with open(".agent-devops.env", "r", encoding="utf-8") as f:
#     text = f.read()
#
# inventory, errors = parse_inventory_from_env(text)
#
# if errors:
#     print("Validation errors:")
#     for e in errors:
#         print(f"- [{e.scope} {e.id}] {e.message}")
#
# print("Servers:", list(inventory.servers.keys()))
